File: gemini_live_sample.py
"""
Gemini Live API サンプル
リクエスト → ツール実行 → ストリーム応答のシンプルな例
"""

import asyncio
import json
import os
import websockets
from datetime import datetime

# Gemini Live API エンドポイント
GEMINI_LIVE_WS_URL = "wss://generativelanguage.googleapis.com/ws/google.ai.generativelanguage.v1alpha.GenerativeService.BidiGenerateContent"


# サンプルツール定義: 現在時刻を取得
def get_current_time():
    """現在時刻を取得するツール"""
    return datetime.now().strftime("%Y-%m-%d %H:%M:%S")


# ツール実行関数
def execute_tool(tool_name, args):
    """ツールを実行して結果を返す"""
    if tool_name == "get_current_time":
        return get_current_time()
    return f"Unknown tool: {tool_name}"


async def run_gemini_live_with_tool():
    """Gemini Live APIでツールを使用した会話を実行"""

    api_key = os.environ.get("GEMINI_API_KEY")
    if not api_key:
        raise ValueError("環境変数 GEMINI_API_KEY を設定してください")

    # APIキーをURLに追加
    url = f"{GEMINI_LIVE_WS_URL}?key={api_key}"

    async with websockets.connect(url) as ws:
        print("✓ WebSocket接続成功\n")

        # 1. セットアップメッセージ（モデル設定 + ツール定義）
        setup_message = {
            "setup": {
                "model": "models/gemini-2.0-flash-exp",
                "generation_config": {
                    "response_modalities": ["TEXT"]
                },
                "tools": [
                    {
                        "function_declarations": [
                            {
                                "name": "get_current_time",
                                "description": "現在の日時を取得します",
                                "parameters": {
                                    "type": "object",
                                    "properties": {}
                                }
                            }
                        ]
                    }
                ]
            }
        }

        await ws.send(json.dumps(setup_message))
        print(f"→ セットアップ送信: {json.dumps(setup_message, ensure_ascii=False, indent=2)}\n")

        # セットアップ応答を受信
        setup_response = await ws.recv()
        print(f"← セットアップ応答: {json.dumps(json.loads(setup_response), ensure_ascii=False, indent=2)}\n")

        # 2. ユーザーメッセージ送信（ツール呼び出しを促すプロンプト）
        user_message = {
            "client_content": {
                "turns": [
                    {
                        "role": "user",
                        "parts": [
                            {
                                "text": "今の時刻を教えてください"
                            }
                        ]
                    }
                ],
                "turn_complete": True
            }
        }

        await ws.send(json.dumps(user_message))
        print(f"→ ユーザーメッセージ送信: {json.dumps(user_message, ensure_ascii=False, indent=2)}\n")

        # 3. ストリーム応答を受信（ツール呼び出しを検出）
        print("=" * 60)
        print("ストリーム応答受信開始")
        print("=" * 60)

        tool_call_detected = False
        tool_call_id = None
        tool_name = None
        tool_args = None

        while True:
            try:
                response = await ws.recv()
                data = json.loads(response)

                print(f"\n← 受信: {json.dumps(data, ensure_ascii=False, indent=2)}")

                # ツール呼び出しの検出
                if "serverContent" in data:
                    server_content = data["serverContent"]

                    if "modelTurn" in server_content:
                        model_turn = server_content["modelTurn"]

                        if "parts" in model_turn:
                            for part in model_turn["parts"]:
                                # Function callの検出
                                if "functionCall" in part:
                                    tool_call_detected = True
                                    function_call = part["functionCall"]
                                    tool_call_id = function_call.get("id")
                                    tool_name = function_call.get("name")
                                    tool_args = function_call.get("args", {})

                                    print(f"\n🔧 ツール呼び出し検出!")
                                    print(f"   ツール名: {tool_name}")
                                    print(f"   ID: {tool_call_id}")
                                    print(f"   引数: {tool_args}")

                                # テキスト応答の表示
                                if "text" in part:
                                    print(f"\n💬 テキスト応答: {part['text']}")

                    # ターン完了チェック
                    if server_content.get("turnComplete"):
                        print("\n✓ ターン完了")
                        break

            except websockets.exceptions.ConnectionClosed:
                print("\n接続が閉じられました")
                break

        # 4. ツールが呼び出された場合、実行して結果を返す
        if tool_call_detected:
            print(f"\n{'=' * 60}")
            print("ツール実行")
            print("=" * 60)

            # ツールを実行
            tool_result = execute_tool(tool_name, tool_args)
            print(f"✓ ツール実行結果: {tool_result}\n")

            # ツール実行結果をGeminiに送信
            tool_response_message = {
                "client_content": {
                    "turns": [
                        {
                            "role": "user",
                            "parts": [
                                {
                                    "functionResponse": {
                                        "id": tool_call_id,
                                        "name": tool_name,
                                        "response": {
                                            "result": tool_result
                                        }
                                    }
                                }
                            ]
                        }
                    ],
                    "turn_complete": True
                }
            }

            await ws.send(json.dumps(tool_response_message))
            print(f"→ ツール実行結果送信: {json.dumps(tool_response_message, ensure_ascii=False, indent=2)}\n")

            # 5. 最終応答をストリームで受信
            print("=" * 60)
            print("最終応答ストリーム受信")
            print("=" * 60)

            full_response = ""

            while True:
                try:
                    response = await ws.recv()
                    data = json.loads(response)

                    print(f"\n← 受信: {json.dumps(data, ensure_ascii=False, indent=2)}")

                    if "serverContent" in data:
                        server_content = data["serverContent"]

                        if "modelTurn" in server_content:
                            model_turn = server_content["modelTurn"]

                            if "parts" in model_turn:
                                for part in model_turn["parts"]:
                                    if "text" in part:
                                        text_chunk = part["text"]
                                        full_response += text_chunk
                                        print(f"\n💬 テキストチャンク: {text_chunk}")

                        if server_content.get("turnComplete"):
                            print("\n✓ ターン完了")
                            break

                except websockets.exceptions.ConnectionClosed:
                    print("\n接続が閉じられました")
                    break

            print(f"\n{'=' * 60}")
            print("完全な応答")
            print("=" * 60)
            print(full_response)


if __name__ == "__main__":
    print("""
╔══════════════════════════════════════════════════════════╗
║     Gemini Live API - ツール実行サンプル                  ║
╚══════════════════════════════════════════════════════════╝
    """)

    try:
        asyncio.run(run_gemini_live_with_tool())
    except Exception as e:
        print(f"\n❌ エラー: {e}")

File: gemini_live_sample.js
/**
 * Gemini Live API サンプル (Node.js版)
 * リクエスト → ツール実行 → ストリーム応答のシンプルな例
 */

const { GoogleGenerativeAI } = require('@google/generative-ai');

// サンプルツール定義: 現在時刻を取得
function getCurrentTime() {
  const now = new Date();
  const year = now.getFullYear();
  const month = String(now.getMonth() + 1).padStart(2, '0');
  const day = String(now.getDate()).padStart(2, '0');
  const hours = String(now.getHours()).padStart(2, '0');
  const minutes = String(now.getMinutes()).padStart(2, '0');
  const seconds = String(now.getSeconds()).padStart(2, '0');

  return `${year}-${month}-${day} ${hours}:${minutes}:${seconds}`;
}

// ツール実行関数
function executeTool(functionCall) {
  const toolName = functionCall.name;

  console.log('\n🔧 ツール呼び出し検出!');
  console.log(`   ツール名: ${toolName}`);
  console.log(`   引数:`, functionCall.args);

  if (toolName === 'get_current_time') {
    const result = getCurrentTime();
    console.log(`   実行結果: ${result}`);
    return result;
  }

  return `Unknown tool: ${toolName}`;
}

async function runGeminiLiveWithTool() {
  const apiKey = process.env.GEMINI_API_KEY;
  if (!apiKey) {
    throw new Error('環境変数 GEMINI_API_KEY を設定してください');
  }

  // 1. クライアント作成
  const genAI = new GoogleGenerativeAI(apiKey);

  // 2. ツール定義
  const tools = [
    {
      functionDeclarations: [
        {
          name: 'get_current_time',
          description: '現在の日時を取得します',
          parameters: {
            type: 'object',
            properties: {}
          }
        }
      ]
    }
  ];

  // 3. モデル取得
  const modelId = 'gemini-2.0-flash-exp';
  console.log(`✓ モデル接続中: ${modelId}\n`);

  const model = genAI.getGenerativeModel({
    model: modelId,
    tools: tools
  });

  // 4. チャットセッション開始
  const chat = model.startChat({
    history: []
  });

  console.log('✓ チャットセッション開始\n');

  // 5. ユーザーメッセージ送信
  const userPrompt = '今の時刻を教えてください';
  console.log('='.repeat(60));
  console.log(`→ ユーザー: ${userPrompt}`);
  console.log('='.repeat(60) + '\n');

  console.log('ストリーム応答受信中...\n');

  // 最初のリクエスト送信（ストリーミング）
  const result = await chat.sendMessageStream(userPrompt);

  let fullText = '';
  const functionCalls = [];

  // 6. ストリーム応答を受信
  for await (const chunk of result.stream) {
    const candidates = chunk.candidates;

    if (candidates && candidates.length > 0) {
      const content = candidates[0].content;

      if (content && content.parts) {
        for (const part of content.parts) {
          // テキスト応答
          if (part.text) {
            console.log(`💬 テキストチャンク: ${part.text}`);
            fullText += part.text;
          }

          // ツール呼び出し
          if (part.functionCall) {
            functionCalls.push(part.functionCall);
          }
        }
      }
    }
  }

  // ストリーム完了を待つ
  const response = await result.response;
  console.log('\n✓ ストリーム完了');

  // 7. ツール実行とレスポンス送信
  if (functionCalls.length > 0) {
    console.log('\n' + '='.repeat(60));
    console.log('ツール実行');
    console.log('='.repeat(60));

    // 各ツールを実行して結果を収集
    const functionResponses = [];

    for (const functionCall of functionCalls) {
      const result = executeTool(functionCall);

      functionResponses.push({
        functionResponse: {
          name: functionCall.name,
          response: {
            result: result
          }
        }
      });
    }

    console.log('\n→ ツール実行結果を送信\n');

    // 8. 最終応答をストリームで受信
    console.log('='.repeat(60));
    console.log('最終応答ストリーム受信中...');
    console.log('='.repeat(60) + '\n');

    const finalResult = await chat.sendMessageStream(functionResponses);

    let finalText = '';

    for await (const chunk of finalResult.stream) {
      const candidates = chunk.candidates;

      if (candidates && candidates.length > 0) {
        const content = candidates[0].content;

        if (content && content.parts) {
          for (const part of content.parts) {
            if (part.text) {
              console.log(`💬 テキストチャンク: ${part.text}`);
              finalText += part.text;
            }
          }
        }
      }
    }

    await finalResult.response;
    console.log('\n✓ ストリーム完了');

    console.log('\n' + '='.repeat(60));
    console.log('完全な応答');
    console.log('='.repeat(60));
    console.log(finalText);
  }
}

// メイン実行
if (require.main === module) {
  console.log(`
╔══════════════════════════════════════════════════════════╗
║     Gemini Live API - ツール実行サンプル (Node.js)       ║
╚══════════════════════════════════════════════════════════╝
  `);

  runGeminiLiveWithTool()
    .then(() => {
      console.log('\n✓ 完了');
      process.exit(0);
    })
    .catch(error => {
      console.error('\n❌ エラー:', error.message);
      process.exit(1);
    });
}

module.exports = { runGeminiLiveWithTool };
